src/com/jdngray77/htmldesigner/backend/BackgroundTask.kt

```diff
@@ -15,23 +22,52 @@ object BackgroundTask : Subscriber {
     }
 
 
+    /**
+     * ### Executes a [runnable] task.
+     *
+     * Executes the given task on a new thread, a pooled thread or the calling thread, at the discretion of the
+     * [threadPool]. There are no guarantees that the task will be executed immediately.
+     */
+    fun execute(runnable: Runnable) {
+        threadPool.execute(runnable)
+    }
+
+
+    @Deprecated("Currently not implemented.", ReplaceWith("TODO()"))
     fun runOnUIThread(runnable: Runnable) {
         TODO()
     }
 
-    
+
     /**
      * Submits a [runnable] task to the for [threadPool] for execution.
+     *
+     * If the thread pool has been instructed to or is already shutdown, the task will not be submitted.
      */
+    @Synchronized
     fun submit(runnable: Runnable) {
-        threadPool.submit(runnable)
+        if (!threadPool.isShutdown)
+            threadPool.submit(runnable)
+    }
+
+
+    /**
+     * Returns all tasks waiting for execution on a worker [Thread] in the [threadPool].
+     *
+     * @return Tasks waiting in the [threadPool] queue for execution.
+     */
+    fun scheduledTasks(): BlockingQueue<Runnable> {
+        val scheduled = threadPool as ThreadPoolExecutor
+
+        return scheduled.queue
     }
 
 
     /**
      * Shutdowns the application's [threadPool].
      *
-     * Executes the [threadPool]'s remaining unfinished tasks and shuts it down on completion.
+     * Rejects the submission of new tasks and executes the [threadPool]'s remaining unfinished tasks - shutting
+     * it down on completion.
      */
     private fun shutdown() {
         threadPool.shutdown()
```

Comment:
This is a mess now lol

Force shutdown then try to gracefully shutdown?

Reply:
Shutdown insturcts an orderly exectution of all remaining tasks in the thread pool, but does not force an immediate shutdow. Await termination then waits for 10 seconds for all tasks to be executed and terminated from the shutdown call. If any tasks remain in the thread pool after 10 seconds, or await termination is interrupted, the thread pool will be forced to shutdown.

That's how I understood it, I will check again.

Reply:
That makes sense!